primitives/core/src/defer.rs

```diff
@@ -66,6 +66,27 @@ macro_rules! defer(
 	};
 );
 
+/// Executes the given code when the current scope is dropped (similarly to [`crate::defer!`]),
+/// unlike [`crate::defer!`] the value of such expression is a guard that can later be cancelled.
+///
+/// # Example
+///
+/// ```rust
+/// let target_file = "to-be-written-atomically.json";
+/// let tmp_file = target_file.with_extension("json.tmp");
+///
+/// // When dropped, `rm_tmp_file_on_drop` will execute the clean up code...
+/// 	let mut rm_tmp_file_on_drop = sp_core::cancellable_defer! {
+/// 		if let Err(reason) = std::fs::remove_file(&tmp_file) {
+/// 			log::error!("Failed to cleanup temp-file: {:?}: {}", tmp_file, reason);
+/// 		}
+/// 	};
+///
+/// std::fs::rename(&tmp_file, target_file).await?;
+///
+/// // ... unless it is cancelled.
+/// rm_tmp_file_on_drop.cancel();
```

Comment:
Won't this fail to compile run as a rustdoc test?

Also, as far as examples go it might be better to pick one without any I/O. (As-is this will leave a junk file after rustdoc runs it.)

Reply:
Fixed.